### PR TITLE
[VL] Support native UDAF

### PR DESCRIPTION
**Pull request by marin-ma**
Add `UserDefinedAggregateFunction` as wrapper for all native udaf.

User should define and register UDAFs in the same way as UDFs. Gluten framework will handle the differences.

issue: #5110

### CODE REVIEW

File: backends-velox/src/main/scala/org/apache/spark/sql/expression/UDFResolver.scala [hunk 1]
```diff
@@ -78,9 +106,12 @@ case class UDFExpression(
 
 object UDFResolver extends Logging {
   private val UDFNames = mutable.HashSet[String]()
-
   private val UDFMap = mutable.HashMap[(String, Seq[DataType]), ExpressionType]()
 
+  private val UDAFNames = mutable.HashSet[String]()
+  private val UDAFMap =
+    mutable.HashMap[(String, Seq[DataType]), (ExpressionType, Seq[AttributeReference])]()
```
Comment (PHILO-HE) [inline backends-velox/src/main/scala/org/apache/spark/sql/expression/UDFResolver.scala:113]: Better to add comments for this map & the above UDFMap.

File: cpp/velox/udf/UdfLoader.cc [hunk 2]
```diff
@@ -77,11 +78,34 @@ std::unordered_set<std::shared_ptr<UdfLoader::UdfSignature>> UdfLoader::getRegis
...
+    return {};
+  }
+  if (!names_.empty()) {
+    return names_;
+  }
+  if (signatures_.empty()) {
+    getRegisteredUdfSignatures();
+  }
+  for (const auto& sig : signatures_) {
+    if (!sig->intermediateType.empty()) {
```
Comment (PHILO-HE) [inline cpp/velox/udf/UdfLoader.cc:104]: Maybe, better to not mix UDF/UDAF signatures together?

**Review by PHILO-HE — APPROVED** [review]
Looks good! Just some suggestions. Please check whether it makes sense (I'm OK to merge this first code version). Thanks!

File: cpp/velox/udf/UdfLoader.cc (hunk 2 above)
Comment (marin-ma) [reply cpp/velox/udf/UdfLoader.cc:104]: Thanks. I've also considered this design. Will address that.

**Pull request merged**